In [1]:
#pip install opencv-python

# **OpenCV Image Processing in Python**
## Overview  
This notebook demonstrates fundamental **image processing** techniques using **OpenCV**.  
It covers **loading, displaying, resizing, converting, and saving images** with practical explanations.

---

##  **What is OpenCV?**
OpenCV (**Open Source Computer Vision Library**) is a powerful library used for **image processing, computer vision, and machine learning** tasks.  

 **Why use OpenCV?**  
-  Efficient image processing  
-  Works with multiple file formats  
-  Supports various transformations like resizing, grayscale conversion, and filtering  

---

## ** Reading and Displaying an Image**  
### **What Happens?**
- The image is loaded into Python memory as a **multi-dimensional array**.
- It is displayed in a separate OpenCV window.
- `cv2.waitKey(2000)` holds the image for **2 seconds** before closing.

### **Why is `cv2.waitKey()` Important?**
- If omitted, the image will **flash and disappear instantly**.
- If no argument (`cv2.waitKey()`), the window remains **open indefinitely**.

---

## **Saving an Image**
### **What Happens?**
- The current image is saved to disk using **JPEG format**.
- OpenCV automatically **compresses the image** when saving.

### **Why Save Images?**
-  To preserve processed images  
-  For further analysis (e.g., AI/ML model training)  
-  To apply filters and transformations later  

---

## **Resizing an Image**
### **What Happens?**
- The image is resized to **500×500 pixels**.
- The aspect ratio is **not preserved**, meaning it may look stretched.

### **Why Resize?**
- Reduces file size  
- Faster processing in ML models  
- Standardizes image dimensions for computer vision tasks  

**Alternative:** Maintain aspect ratio by setting **one dimension** and letting OpenCV adjust the other.

---

## **Converting to Grayscale**
### **What Happens?**
- OpenCV removes color channels (RGB) and keeps **intensity values**.
- This simplifies the image to **one channel** (shades of gray).

### **Why Convert to Grayscale?**
- Reduces complexity  
- Speeds up processing  
- Essential for edge detection and pattern recognition  

**Example Usage:**  
- Face detection  
- Object recognition  
- Image segmentation  

---

## **Understanding Image Dimensions**
### **What Happens?**
- `img.shape[0]`: Gets the **height** of the image.
- `img.shape[1]`: Gets the **width** of the image.
- Multiplying by `0.5` calculates **half the size**.

### **Why Do We Need Image Dimensions?**
- Helps in resizing  
- Useful for cropping and transformations  
- Needed for AI models that expect fixed input sizes  

---

## **Summary of Concepts**
| Feature             | What Happens?  | Why is it Useful?  |
|--------------------|---------------|-------------------|
| **Read Image**      | Loads an image into memory | Enables processing |
| **Display Image**   | Opens a window to show the image | Visual verification |
| **Save Image**      | Saves image as a new file | Stores processed results |
| **Resize Image**    | Changes image dimensions | Optimizes performance |
| **Grayscale Conversion** | Removes color, keeps intensity values | Simplifies processing |
| **Image Shape**     | Retrieves width & height | Useful for transformations |

---

## **Final Thoughts**
- **OpenCV is powerful!** Even with basic operations, it enables **complex applications**.
- **Practical Use Cases:**
  - Object & face detection  
  - Image classification  
  - Feature extraction  


## **Import Required Libraries**

In [11]:
import cv2
import numpy as np

**Read and Display the Image**

In [12]:
img=cv2.imread("Modi.jpg",1)

In [13]:
# print(img)        # matrix 3D - this is how images are stored 

In [14]:
img.shape

(1358, 1500, 3)

- shape gives the dimensions of the image array 
- The 3D are in 1358 pixels in height * 1500 pixels wide
- 3 means that there are 3 components (RGB) that make up this image

## **Display the image**

- The first parameter will be title shown on image window
- The second parameter is teh image variable

- By leaving wait key blank, it just waits for anykey to be pressed before continuing
- By placing number ( expect 0), we can specify a delay for how long you keep window open (time is in  milliseconds here )

- destroy all window is used to close all open windows
- Failure to place this will cause your programe to hang

In [15]:
cv2.imshow('prime minister',img)
cv2.waitKey(2000)           # if nothing in waitKey , runs for infinite times
cv2.destroyAllWindows()

## **Save the image**

imwrite specifing the file name and the image to be saved

In [16]:
cv2.imwrite("saving_modi.jpg",img)

True

## **Resize the image**

In [17]:
img=cv2.imread("Modi.jpg")
resized_img=cv2.resize(img,(500,500))


## **Convert to Grayscale and Display**

In [ ]:
gray=cv2.cvtColor(resized_img,cv2.COLOR_BGR2GRAY)
cv2.imshow('modi image',gray)
cv2.waitKey()          
cv2.destroyAllWindows()

## **Image Dimensions**

In [20]:
new_height = int(img.shape[0] * 0.5)
new_width = int(img.shape[1] * 0.5)
resized_img = cv2.resize(img, (new_width, new_height))
img

array([[[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       ...,

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]],

       [[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]]

In [9]:
img.shape[0]*0.5

679.0

In [10]:
img.shape[1]*0.5

750.0

### **Project by : SIREESHA RAGIPATI**